heap.ts

```diff
@@ -119,6 +119,165 @@ export class BumpAllocator implements MarkableAllocator {
 //   }
 // BitMappedBlocks: [infomap, bucket1, bucket2, bucket3...]
 
+export class BitMappedBlocks implements MarkableAllocator {
+  start: bigint;
+  end: bigint;
+  numBlocks: bigint;
+  blockSize: bigint;
+  infomap: Uint8Array;
+
+  constructor(start: bigint, end: bigint, blockSize: bigint) {
+    this.start = start + (start % 2n); // Align at even boundary
+    this.end = end;
+    this.blockSize = blockSize;
+    
+    this.numBlocks = (end - start)/blockSize;
+
+    // one byte for free/used, n bytes for header
+    const totalNBytes = Number((1n + BigInt(HEADER_SIZE_BYTES)) * this.numBlocks)
+    this.infomap = new Uint8Array(totalNBytes);
+  }
+
+  getInfomapIndex (index: bigint): number {
+    return Number((1n + BigInt(HEADER_SIZE_BYTES)) * index)
+  }
+
+  isFree (index: bigint): boolean {
+    return this.infomap[this.getInfomapIndex(index)] === 0
+  }
+
+  getNumFreeBlocks() {
+    let count = 0;
+    for(let index = 0n; index < this.numBlocks; ++index) {
+      if(this.isFree(index)) {
+        ++count;
+      }
+    }
+
+    return count;
+  }
+
+  // Returns the block index that can satisfy the requested `size`
+  // Returns -1 to indicate a failure
+  getBlockIndex (size: bigint): bigint {
+    // How many blocks are needed to satisfy the request
+    const blocksRequested = size/this.blockSize;
+
+    // Linearly Traverse the infomap to find blocks
+    // This can lead to fragmentation
+    // Can this be optimized further?
+    let index = 0n;
+    while(index < this.numBlocks) {
+
+      // Find the first free block
+      while(!this.isFree(index)) {
+        ++index;
+      }
+
+      // Hit the end
+      if(index + blocksRequested > this.numBlocks) {
+        return -1n;
+      }
+
+      let count = 1n; // Keep track of free blocks
+
+      while(count < blocksRequested && index < this.numBlocks) {
+        ++count;
+        ++index;
+      }
+
+      if(count == blocksRequested) {
+        return index;
+      }
+    }
+
+    return -1n;
+  }
+
+  alloc(size: bigint): Block {
+    // Search for a free block(or a group of contiguous free blocks)`
+    const blockIndex = this.getBlockIndex(size);
+    if(blockIndex === -1n) {
+      return NULL_BLOCK;
+    }
+
+    let nBlocks = size/this.blockSize;
+    if(size % this.blockSize !== 0n) {
+      nBlocks += 1n; // (Hack) in place of Math.ceil
+    }
+
+    // Set the bit(byte) as "used"
+    // Size is stored in header -  Useful when sweeping
+    for(let index = blockIndex; index < blockIndex + nBlocks; ++index) {
+      this.infomap[Number((1n + BigInt(HEADER_SIZE_BYTES))*index)] = 1;
+    }
+
+    return {
+      ptr: this.start + BigInt(blockIndex) * this.blockSize,
+      size: size
+    }
+  }
+
+  free2(ptr: bigint) {
+    // mark "ptr" block as free
+    const index = (ptr - this.start)/this.blockSize;
+    this.infomap[this.getInfomapIndex(index)] = 0
+  }
```

Comment:
With multi-block allocations, this will cause memory to leak because there's no way to determine the size from the pointer and we don't store that metadata.

This shows the original motivation from the CppCon2015 talk (free(block: Block)) which contains pointer + size information.

We have some options:

Swap to the free(block: Block) implementation
Store the metadata in the info byte (so it no longer just means "allocated" but also the "number of contiguous blocks" to be freed together)
Throw an error on allocation requests bigger than the block size

Option 2 seems like a feasible option and a simple update (and I think it's reasonable to through an error if we request more than 256 blocks of a given size).